File: core/src/androidTest/java/com/emarsys/core/resource/MetaDataReaderTest.java
package com.emarsys.core.resource;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.os.Bundle;
import androidx.test.InstrumentationRegistry;

import com.emarsys.testUtil.TimeoutUtils;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestRule;
import org.mockito.Mockito;

import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.nullable;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class MetaDataReaderTest {

    private MetaDataReader reader;

    @Rule
    public TestRule timeout = TimeoutUtils.getTimeoutRule();

    @Before
    public void setUp() throws Exception {
        reader = new MetaDataReader();
    }

    @Test(expected = IllegalArgumentException.class)
    public void testGetIntOrNull_context_mustNotBeNull() throws Exception {
        reader.getInt(null, "key");
    }

    @Test(expected = IllegalArgumentException.class)
    public void testGetIntOrNull_key_mustNotBeNull() throws Exception {
        reader.getInt(InstrumentationRegistry.getContext(), null);
    }

    @Test
    public void testGetIntOrNull_returnsValue_ifExists() throws PackageManager.NameNotFoundException {
        ApplicationInfo applicationInfo = new ApplicationInfo();
        Bundle bundle = new Bundle();
        bundle.putInt("something", 42);
        applicationInfo.metaData = bundle;

        Context context = mock(Context.class, Mockito.RETURNS_DEEP_STUBS);
        when(context.getPackageManager().getApplicationInfo(nullable(String.class), anyInt())).thenReturn(applicationInfo);

        assertEquals(42, reader.getInt(context, "something"));
    }

    @Test
    public void getIntOrNull_shouldReturnNull_ifThereIsNoValue() throws PackageManager.NameNotFoundException {
        ApplicationInfo applicationInfo = new ApplicationInfo();
        applicationInfo.metaData = new Bundle();

        Context context = mock(Context.class, Mockito.RETURNS_DEEP_STUBS);
        when(context.getPackageManager().getApplicationInfo(nullable(String.class), anyInt())).thenReturn(applicationInfo);

        assertEquals(0, reader.getInt(context, "something"));
    }

    @Test(expected = IllegalArgumentException.class)
    public void testGetInt_context_mustNotBeNull() throws Exception {
        reader.getInt(null, "key", 0);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testGetInt_key_mustNotBeNull() throws Exception {
        reader.getInt(InstrumentationRegistry.getContext(), null, 0);
    }

    @Test
    public void testGetInt_returnsValue_ifExists() throws PackageManager.NameNotFoundException {
        ApplicationInfo applicationInfo = new ApplicationInfo();
        Bundle bundle = new Bundle();
        bundle.putInt("something", 43);
        applicationInfo.metaData = bundle;

        Context context = mock(Context.class, Mockito.RETURNS_DEEP_STUBS);
        when(context.getPackageManager().getApplicationInfo(nullable(String.class), anyInt())).thenReturn(applicationInfo);

        assertEquals(43, reader.getInt(context, "something", -1));
    }

    @Test
    public void getInt_shouldReturnDefaultValue_ifThereIsNoValue() throws PackageManager.NameNotFoundException {
        ApplicationInfo applicationInfo = new ApplicationInfo();
        applicationInfo.metaData = new Bundle();

        Context context = mock(Context.class, Mockito.RETURNS_DEEP_STUBS);
        when(context.getPackageManager().getApplicationInfo(nullable(String.class), anyInt())).thenReturn(applicationInfo);

        assertEquals(0, reader.getInt(context, "something"));
        assertEquals(200, reader.getInt(context, "something", 200));
    }
}

File: core/src/androidTest/java/com/emarsys/core/resource/MetaDataReaderTest.kt
package com.emarsys.core.resource

import android.content.Context
import android.content.pm.ApplicationInfo
import android.os.Bundle
import com.emarsys.testUtil.InstrumentationRegistry.Companion.getTargetContext
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import org.junit.Before
import org.junit.Test

class MetaDataReaderTest {
    private var reader: MetaDataReader? = null
    private lateinit var mockContext: Context
    private lateinit var applicationInfo: ApplicationInfo

    @Before
    @Throws(Exception::class)
    fun setUp() {
        applicationInfo = ApplicationInfo()
        mockContext = mockk(relaxed = true)
        every {
            mockContext.packageManager.getApplicationInfo(
                any(), any<Int>()
            )
        } returns applicationInfo

        reader = MetaDataReader()
    }

    @Test
    fun testGetIntOrNull_context_mustNotBeNull() {
        shouldThrow<IllegalArgumentException> {
            reader!!.getInt(null, "key")
        }
    }

    @Test
    fun testGetIntOrNull_key_mustNotBeNull() {
        shouldThrow<IllegalArgumentException> {
            reader!!.getInt(getTargetContext(), null)
        }
    }

    @Test
    fun testGetIntOrNull_returnsValue_ifExists() {
        val bundle = Bundle()
        bundle.putInt("something", 42)
        applicationInfo.metaData = bundle

        reader!!.getInt(mockContext, "something").toLong() shouldBe 42
    }

    @Test
    fun intOrNull_shouldReturnNull_ifThereIsNoValue() {
        applicationInfo.metaData = Bundle()

        reader!!.getInt(mockContext, "something").toLong() shouldBe 0
    }

    @Test
    fun testGetInt_context_mustNotBeNull() {
        shouldThrow<IllegalArgumentException> {
            reader!!.getInt(null, "key", 0)
        }
    }

    @Test
    fun testGetInt_key_mustNotBeNull() {
        shouldThrow<IllegalArgumentException> {
            reader!!.getInt(getTargetContext(), null, 0)
        }
    }

    @Test
    fun testGetInt_returnsValue_ifExists() {
        val bundle = Bundle()
        bundle.putInt("something", 43)
        applicationInfo.metaData = bundle

        reader!!.getInt(mockContext, "something", -1).toLong() shouldBe 43
    }

    @Test
    fun int_shouldReturnDefaultValue_ifThereIsNoValue() {
        applicationInfo.metaData = Bundle()
        reader!!.getInt(mockContext, "something").toLong() shouldBe 0
        reader!!.getInt(mockContext, "something", 200).toLong() shouldBe 200
    }
}